### PR TITLE
I removed ReadOnlySequence common tests from factory tests

### PR DESCRIPTION
**Pull request by AlexRadch**
I removed ReadOnlySequence common tests from factory tests. They are duplicates of comment tests and they does not use factory.

### CODE REVIEW

**Comment by ahsonkhan** [comment]
Can you please verify that the code coverage numbers don't change when you remove them (before/after)?

**Comment by AlexRadch** [comment]
> Can you please verify that the code coverage numbers don't change when you remove them (before/after)?

Good question. How to make this?

**Comment by ahsonkhan** [comment]
> cd src\System.Memory\tests\
msbuild /t:BuildAndTest /p:Coverage=true

See this document for more details: https://github.com/dotnet/corefx/blob/master/Documentation/building/code-coverage.md

**Comment by AlexRadch** [comment]
Code coverage before:
```
  === TEST EXECUTION SUMMARY ===
     System.Memory.Tests  Total: 42541, Errors: 0, Failed: 0, Skipped: 0, Time: 22,566s
  Committing...
  Visited Classes 27 of 28 (96.43)
  Visited Methods 328 of 349 (93.98)
  Visited Points 5745 of 6082 (94.46)
  Visited Branches 2393 of 2566 (93.26)
  
  ==== Alternative Results (includes all methods including those without corresponding source) ====
  Alternative Visited Classes 28 of 30 (93.33)
  Alternative Visited Methods 332 of 353 (94.05)
  ----- end  0:57:45,42 ----- exit code 0 ----------------------------------------------------------
```
Code coverage after:
```
  === TEST EXECUTION SUMMARY ===
     System.Memory.Tests  Total: 42453, Errors: 0, Failed: 0, Skipped: 0, Time: 22,141s
  Committing...
  Visited Classes 27 of 28 (96.43)
  Visited Methods 328 of 349 (93.98)
  Visited Points 5745 of 6082 (94.46)
  Visited Branches 2393 of 2566 (93.26)
  
  ==== Alternative Results (includes all methods including those without corresponding source) ====
  Alternative Visited Classes 28 of 30 (93.33)
  Alternative Visited Methods 332 of 353 (94.05)
  ----- end  0:59:50,38 ----- exit code 0 ----------------------------------------------------------
```

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.char.cs [hunk 1]
```diff
@@ -133,108 +133,6 @@ public void ReadOnlyBufferSlice_ChecksEnd()
             Assert.Throws<ArgumentOutOfRangeException>(() => buffer.Slice(70, buffer.Start));
         }
 
-        [Fact]
-        public void SegmentStartIsConsideredInBoundsCheck()
```
Comment (ahsonkhan) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.char.cs]: The common test only contains the test for T = byte, but this is testing T = char (https://github.com/dotnet/corefx/blob/master/src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.cs#L18).

I think we should keep the tests in general. No need to remove tests at this point.

**Comment by AlexRadch** [comment]
Those tests should not be in factory tests. They does not use factories and they does the same test for each subclass
```
        public class Array : ReadOnlySequenceTestsByte
        public class Memory : ReadOnlySequenceTestsByte
        public class SingleSegment : ReadOnlySequenceTestsByte
        public class SegmentPerByte : ReadOnlySequenceTestsByte
        public class SplitInThreeSegments : ReadOnlySequenceTestsByte
```
and one more times in common tests. I can add common tests for char type.

**Comment by ahsonkhan** [comment]
> one more times in common tests. I can add common tests for char type.
We run the same tests 6 times.

OK, that will work. Thanks you.

**Comment by AlexRadch** [comment]
I created common tests for byte and char types.

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceFactory.byte.cs [hunk 2]
```diff
@@ -103,32 +102,25 @@ public override ReadOnlySequence<T> CreateWithContent(T[] data)
...
-
-                for (int j = 0; j < length; j++)
-                {
-                    chars[dataOffset + j] = s[j];
-                }
-
-                Memory<T> memory = new Memory<T>(chars).Slice(length, length);
-
+                int length = input.Length;
+                int dataOffset = length / 2;
```
Comment (ahsonkhan) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceFactory.byte.cs]: **Question:** This used to be `dataOffset = length`, why change it to `length / 2`?

File: src/System.Memory/src/System/Buffers/ReadOnlySequence.cs [hunk 3]
```diff
@@ -85,6 +85,7 @@ public ReadOnlySequence(ReadOnlySequenceSegment<T> startSegment, int startIndex,
         {
             if (startSegment == null ||
                 endSegment == null ||
+                (startSegment != endSegment && startSegment.RunningIndex > endSegment.RunningIndex) ||
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Buffers/ReadOnlySequence.cs]: @pakrym, is this check valid?

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.byte.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            const int blockSize = 4096;
+
+            byte[] items = Encoding.ASCII.GetBytes("Hello World");
+            byte[] firstItems = Enumerable.Repeat((byte)'a', blockSize - 5).Concat(items.Take(5)).ToArray();
+            byte[] secondItems = items.Skip(5).Concat(Enumerable.Repeat((byte)'a', blockSize - (items.Length - 5))).ToArray();
+
+            BufferSegment<byte> firstSegment = new BufferSegment<byte>(firstItems);
+            BufferSegment<byte> secondSegment = firstSegment.Append(secondItems);
+
+            ReadOnlySequence<byte> buffer = new ReadOnlySequence<byte>(firstSegment, 0, secondSegment, items.Length - 5);
```
Comment (ahsonkhan) [inline src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceTests.Common.byte.cs]: nit: use var

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceFactory.byte.cs (hunk 2 above)
Comment (AlexRadch) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceFactory.byte.cs]: Would it be better to move tested to middle of array? Or stay them at end?

File: src/System.Memory/src/System/Buffers/ReadOnlySequence.cs (hunk 3 above)
Comment (pakrym) [reply src/System.Memory/src/System/Buffers/ReadOnlySequence.cs]: Yes. But with this amount of checks Pipelines would soon need a new unsafe ctor.

File: src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceFactory.byte.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Memory/tests/ReadOnlyBuffer/ReadOnlySequenceFactory.byte.cs]: It is fine as is. Thanks.

**Comment by ahsonkhan** [comment]
@dotnet-bot test Windows x64 Debug Build

@pakrym, good to merge?

**Comment by pakrym** [comment]
LGTM

**Pull request merged**